File: ruby/bowling/bowling.rb
class Game
  attr_reader :frames

  def initialize
    @frames = []
  end

  def roll(pins)
    validate_roll(pins, frames[-1])

    if frame_in_progress?
      frames[-1].push(pins)
    else
      frames.push([pins])
    end
  end

  def score
    validate_game

    frames.each_with_index.inject(0) do |total, (frame, index)|
      if is_bonus_frame?(index)
        total
      elsif is_spare?(frame)
        total + frame.inject(0, :+) + next_roll_value(index)
      elsif is_strike?(frame)
        total + frame.inject(0, :+) + next_2_rolls_value(index)
      else
        total + frame.inject(0, :+)
      end
    end
  end

  private

  def frame_in_progress?
    frames[-1] &&
    frames[-1].inject(0, :+) < 10 && 
    frames[-1].length < 2
  end

  def is_spare?(frame)
    frame.length == 2 && frame.inject(0, :+) === 10
  end

  def is_strike?(frame)
    frame[0] === 10
  end

  def next_roll_value(from)
    next_frame = frames[from + 1]
    next_frame[0]
  end

  def next_2_rolls_value(from)
    next_two_frames = frames[(from + 1)..(from + 2)]
    next_several_rolls = next_two_frames.flatten
    next_several_rolls[0] + next_several_rolls[1]
  end

  def is_bonus_frame?(frame_index)
    frame_index >= 10
  end

  def complete_game?
    complete_with_open_frame? ||
    complete_with_spare_bonus? ||
    complete_with_strike_bonus?
  end

  def complete_with_open_frame?
    frames.length == 10 &&
    frames[-1].length == 2 &&
    frames[-1].inject(:+) < 10
  end

  def complete_with_spare_bonus?
    frames[9] &&
    frames[9].length == 2 &&
    frames[9].inject(0, :+) == 10 &&
    frames[10] &&
    frames[10].length == 1
  end

  def complete_with_strike_bonus?
    frames[9] &&
    frames[9].length == 1 &&
    frames[9][0] == 10 &&
    frames[10] &&
    (frames[10].length == 2 ||
     (frames[10][0] == 10 &&
      frames[11] &&
      frames[11].length == 1
     )
    )
  end

  def validate_roll(pins, currentFrame)
    raise BowlingError if pins < 0 || pins > 10
    if frame_in_progress?
      raise BowlingError if pins + currentFrame[0] > 10
    end
    raise BowlingError if complete_game?
  end

  def validate_game
    raise BowlingError if frames.length < 10
    raise BowlingError if !complete_game?
  end

  class BowlingError < StandardError
  end
end

File: typescript/bowling/bowling.ts
type Frame = number[];

export class Bowling {
  frames: Frame[] = [];

  public roll(pins: number): void {
    if (this.gameOver()) throw new Error("Cannot roll after game is over");

    let isFinalFrame = this.frames.length === 10;
    let currentFrame = this.frames[this.frames.length - 1];
    let isMidFrame =
      this.frames.length > 0 && currentFrame.length < 2 && currentFrame[0] < 10;

    if (isMidFrame || isFinalFrame) {
      currentFrame.push(pins);
    } else {
      this.frames.push([pins]);
    }

    if (this.tooFewPins(pins)) throw new Error("Negative roll is invalid");
    if (this.tooManyPins(pins))
      throw new Error("Pin count exceeds pins on the lane");
    if (this.invalidRoll())
      throw new Error("Pin count exceeds pins on the lane");
  }

  public score(): number {
    if (!this.gameStarted() || !this.gameOver())
      throw new Error("Score cannot be taken until the end of the game");

    return this.frames.reduce((total, frame, i) => {
      return total + this.getFrameTotal(frame, i);
    }, 0);
  }

  private getFrameTotal(frame: Frame, frameIndex: number): number {
    return frame.reduce((frameTotal, roll, j) => {
      let isFinalFrame = frameIndex === 9;
      let isSpare = frame.length === 2 && frame[0] + frame[1] === 10 && j === 1;
      let isStrike = frame.length === 1 && frame[0] === 10;

      if (isSpare && !isFinalFrame) {
        return frameTotal + roll + this.getSpareBonus(frameIndex);
      } else if (isStrike && !isFinalFrame) {
        return frameTotal + roll + this.getStrikeBonus(frameIndex);
      } else {
        return frameTotal + roll;
      }
    }, 0);
  }

  private getSpareBonus(frameIndex: number): number {
    return this.frames[frameIndex + 1][0];
  }

  private getStrikeBonus(frameIndex: number): number {
    let isPenultimateFrame = frameIndex === 8;
    let firstRoll = this.frames[frameIndex + 1][0];
    let secondRoll;

    if (firstRoll === 10 && !isPenultimateFrame) {
      secondRoll = this.frames[frameIndex + 2][0];
    } else {
      secondRoll = this.frames[frameIndex + 1][1];
    }

    return firstRoll + secondRoll;
  }

  /*--- Validation Methods ----*/

  private tooFewPins(pins: number): boolean {
    return pins < 0;
  }

  private tooManyPins(pins: number): boolean {
    return pins > 10;
  }

  private invalidRoll(): boolean {
    let inFinalFrame = this.frames.length === 10;

    return inFinalFrame
      ? this.invalidFinalFrame()
      : this.invalidStandardFrame();
  }

  private invalidStandardFrame(): boolean {
    let lastFrame = this.frames[this.frames.length - 1];
    return lastFrame.reduce((sum, num) => sum + num) > 10;
  }

  private invalidFinalFrame(): boolean {
    let lastFrame = this.frames[this.frames.length - 1];
    let total = lastFrame.reduce((sum, num) => sum + num);

    return total > 30 || (lastFrame[1] < 10 && total > 20);
  }

  private gameStarted(): boolean {
    return this.frames.length > 0;
  }

  private gameOver(): boolean {
    let isFinalFrame = this.frames.length === 10;
    if (!isFinalFrame) return false;

    let frame = this.frames[this.frames.length - 1];
    let hasBonus = frame[0] + frame[1] >= 10;

    return (
      (hasBonus && frame.length === 3) || (!hasBonus && frame.length === 2)
    );
  }
}
